cupyx/_pinned_array.py

```diff
@@ -0,0 +1,157 @@
+import numpy
+
+from cupy import cuda
+from cupy._creation.basic import _new_like_order_and_strides
+from cupy.core import internal
+
+
+def _update_shape(a, shape):
+    if shape is None and a is not None:
+        shape = a.shape
+    elif isinstance(shape, int):
+        shape = (shape,)
+    else:
+        shape = tuple(shape)
+    return shape
+
+
+def _update_order_char(x, order_char):
+    # This is a pure Python version of cupy.core.core._update_order_char()
```

Comment:
We could relax the typing in `cupy.core.core._update_order_char()` so that it can be used here. However, doing so increases the overhead a little bit, as we have to access Python attributes (`x.flags.f_contiguous`) instead of C attributes (`x._f_contiguous`) which is absent in `numpy.ndarray`:
|  | time |
|--- | --- |
| original `cupy.core.core._update_order_char()` | 0.13 us |
| remove the `ndarray` typing from `x` (but still kept in Cython) | 0.48 us |
| full Python implementation (as done above) | 0.90 us |

Let me know if this overhead is acceptable for the purpose of reducing code duplication.

Reply:
How about changing the original `_update_order_char` function to take `bool c_contiguous, bool f_continguous`  instead of `ndarray`?

Reply:
Good point, @kmaehashi! I should have thought about it...Done in b833b24. No performance regression is observed, with a caveat that in Python the (likely small) overhead is swept into function parameter preparation (`c_contiguous` and `f_contiguous`).